File: kotlin/subprojects/numbers-utils/src/test/kotlin/com/prog/numbers/utils/operations/FactorialUtilsTests.kt
package com.prog.numbers.utils.operations

import org.apache.commons.lang3.time.DateFormatUtils
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FactorialUtilsTests {
    /**
     * # `formatErrorMessage`
     *
     * This function formats the error message for the test.
     *
     * __Arguments:__
     *
     * * `expected: BigInteger` - The expected output.
     * * `actual: BigInteger` - The actual output.
     *
     * __Return:__
     *
     * * `String` - The formatted error message.
     */
    private fun formatErrorMessage(
        expected: BigInteger,
        actual: BigInteger,
    ): String {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "yyyy-MM-dd'T'HH:mm:ss")
        return """
            |${"\u001B[31m"}
            |Test run at: $formattedDate
            |Expected : $expected
            |  Actual : $actual
            |
            |${"\u001B[00m"}
            """.trimMargin()
    }

    /**
     * # `factorial test`
     *
     * This function tests the `factorial` function.
     */
    @Test
    fun `factorial test`() {
        val n = 75
        val expectedOutput =
            BigInteger(
                "24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000",
            )
        val actualOutput = FactorialUtils.factorial(n)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `factorial random test`
     *
     * This function tests the `factorialRandom` function.
     */
    @Test
    fun `factorial random test`() {
        val result = FactorialUtils.factorialRandom()
        val startString = "Fact["
        val errorMessage =
            """
            |${"\u001B[31m"}
            |Expected: $startString
            |     Got: $result${"\u001B[00m"}
            """.trimMargin()
        assertTrue(result.startsWith(startString), errorMessage)
        assertTrue(result.contains(" = "))
    }
}

File: rust/crates/numbers-utils/tests/operations/factorial_utils_tests.rs
use chrono::Local;
use colored::Colorize;

use num_bigint::BigUint;
use numbers_utils::operations::factorial_utils::{factorial, factorial_random};

/// # `format_error_message`
///
/// This function formats the error message for the test cases.
///
/// __Arguments:__
///
/// * `expected: BigUint` - The expected output.
/// * `actual: BigUint` - The actual output.
///
/// __Return:__
///
/// * `colored::ColoredString` - The formatted error message.
fn format_error_message(expected: &BigUint, actual: &BigUint) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

/// # `factorial_test`
///
/// This function tests the `factorial` function.
#[test]
fn factorial_test() {
    let n = 75;
    let expected_output = BigUint::parse_bytes(b"24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000", 10).unwrap();
    let actual_output = factorial(n);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `factorial_random_test`
///
/// This function tests the `factorial_random` function.
#[test]
fn factor_random_test() {
    let result = factorial_random();
    let start_string = "Fact[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}
